crates/host/src/cache.rs

```diff
@@ -0,0 +1,581 @@
+    time::Instant,
+};
+use tokio::sync::RwLock;
+use tokio::time::Duration;
+use tokio_util::time::{delay_queue, DelayQueue};
+
+#[derive(Debug, Clone)]
+struct CacheEntry<V>
+where
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    value: V,
+    created_at: Instant,
+}
+
+impl<V> CacheEntry<V>
+where
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn new(value: V) -> Self {
+        Self {
+            value,
+            created_at: Instant::now(),
+        }
+    }
+
+    fn get_value(&self) -> V {
+        self.value.clone()
+    }
+
+    fn get_created_at(&self) -> Instant {
+        self.created_at
+    }
+}
+
+#[derive(Debug)]
+struct CacheInternal<K, V>
+where
+    K: std::hash::Hash + Eq + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    data: HashMap<K, (CacheEntry<V>, delay_queue::Key)>,
+    expirations: DelayQueue<K>,
+    ttl: Duration,
+    lazy: bool,
+}
+
+impl<K, V> CacheInternal<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    fn new(ttl: Duration) -> Self {
+        Self {
+            data: HashMap::new(),
+            expirations: DelayQueue::new(),
+            ttl,
+            lazy: true,
+        }
+    }
+
+    fn insert(&mut self, key: K, value: V) {
+        let delay = self.expirations.insert(key.clone(), self.ttl);
+        self.data
+            .insert(key.clone(), (CacheEntry::new(value), delay));
+    }
+
+    fn get(&mut self, key: &K) -> Option<V> {
+        match self.data.entry(key.clone()) {
+            std::collections::hash_map::Entry::Occupied(entry)
+                if entry.get().0.get_created_at().elapsed() < self.ttl =>
+            {
+                Some(entry.get().0.get_value())
+            }
+            std::collections::hash_map::Entry::Occupied(entry) => {
+                if self.lazy {
+                    let (_, delay_key) = entry.remove();
+                    self.expirations.remove(&delay_key);
+                }
+                None
+            }
+            std::collections::hash_map::Entry::Vacant(_) => None,
+        }
+    }
+
+    fn remove(&mut self, key: &K) {
+        if let Some((_, cache_key)) = self.data.remove(key) {
+            self.expirations.remove(&cache_key);
+        }
+    }
+
+    fn poll_purge(&mut self, cx: &mut Context<'_>) -> Poll<()> {
+        while let Some(entry) = ready!(self.expirations.poll_expired(cx)) {
+            self.data.remove(entry.get_ref());
+        }
+
+        Poll::Ready(())
+    }
+}
+
+/// A thread-safe, TTL-based cache with configurable purging strategies
+pub struct Cache<K, V>
+where
+    K: std::hash::Hash + Eq + Clone + std::fmt::Debug + Send + Sync + 'static,
+    V: std::fmt::Debug + Clone + Send + Sync + 'static,
+{
+    internal: Arc<RwLock<CacheInternal<K, V>>>,
+    purge_cycles: Arc<Option<u32>>,
```

Comment:
Sorry @JeancarloBarrios, I think my comment wasn't clear. This line doesn't need to be `Arc` as copying a u32 is pretty much nothing. I was actually talking about changing `purge_cycle_count` to be `Arc<AtomicU32>`. That way if a cache is cloned, the purge cycle count stays the same across all cloned caches

Reply:
No, you were clear. Idk what I was thinking. I was distracted.  @thomastaylor312